[RFC]: File Manager documents

pazos opened the pull request:
Related to https://github.com/koreader/koreader/issues/9575

## NonDocument

A generic handler for files that are not documents.

Purpose: allow plugins to register specific actions to open files by usual means (from FM, from dialogs, drag&drop ...)

Sample usage is included in the `hello.koplugin` 

### to do
- naming convention
- review documentation


<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/koreader/koreader/9615)
<!-- Reviewable:end -->

NiLuJe commented on frontend/document/filemanagerdocument.lua:
```diff
@@ -0,0 +1,85 @@
+local Document = require("document/document")
+local util = require("util")
+
+local FileManagerDocument = Document:new{
```
`extend` ;p

NiLuJe commented on frontend/document/filemanagerdocument.lua:
```diff
@@ -0,0 +1,85 @@
+    _document = false,
+    provider = "filemanagerdocument",
+    provider_name = "File Manager",
+    providers = {},
```
This will need some thinking, unless addHandler takes care or don't inserting the same provider twice, or if the implementation provides a way for Plugins to deregister their handler on teardown.

(This is a static class members, while plugins will call init on each FileManager/ReaderUI instantiation, so, right now, you get a bunch of duplicates and pinned references).

NiLuJe replied:
But the cached approach is complicated by the fact that if the plugin registers an instance reference instead of a module-local, this'll pin the instance object, too.

NiLuJe replied:
And DocumentRegistry has the same design issue is that's it's very much designed only for static handlers.

pazos replied:
I think I solved it. Now providers is an array with names of providers.
Also we just register an extension once, when the first handler for the extension is added. More than one handler for a extension shouldn't reach documentRegistry. Instead we show a menu to let the user choose between actions.

NiLuJe replied:
Yup, that looks better at a glance ;).

Docs should make it clear that `open_func` should point to a static/class method and not an instance one (which shouldn't be a problem given the design; and holds in the example), though ;).

NiLuJe commented:
The core changes don't look too bad ;).

NiLuJe commented on plugins/shell-runner.koplugin/main.lua:
```diff
@@ -0,0 +1,61 @@
+    }
+end
+
+local ShellRunner = WidgetContainer:new{
```
Also `extend` ;).

NiLuJe commented on plugins/shell-runner.koplugin/main.lua:
```diff
@@ -0,0 +1,61 @@
+
+-- on init we register the extensions as FM documents
+function ShellRunner:init()
+    FileManagerDocument:addHandler("ShellRunner", t)
```
Could possibly do nothing in ReaderUI?

e.g., `if self.ui.name == "ReaderUI" then return end`

pazos replied:
is this still relevant?

poire-z commented on frontend/apps/filemanager/filemanager.lua:
```diff
@@ -797,7 +803,7 @@ function FileManager:tapPlus()
     UIManager:show(self.file_dialog)
 end
 
-function FileManager:reinit(path, focused_file)
+function FileManager:reinit(path, focused_file, callback)
```
Not fond of seeing this additional parameter because of this "little" feature :/
At lease, give it a less anonymous and really explicite name so we can ignore it when reading this code.
`filemanagerdocument_handle_callback` ?

poire-z commented on frontend/apps/filemanager/filemanager.lua:
```diff
@@ -819,6 +825,7 @@ function FileManager:reinit(path, focused_file)
     if self.select_mode then
         self.title_bar:setRightIcon("check")
     end
+    self:handleFile(focused_file, callback)
```
Same expliciteness required : `self:handleFileManagerDocument(focused_file, filemanagerdocument_handle_callback )`
(Given the edgy context, I don't mind this being long :)

poire-z commented on frontend/apps/reader/readerui.lua:
```diff
@@ -14,6 +14,7 @@ local DocumentRegistry = require("document/documentregistry")
 local Event = require("ui/event")
 local FileManagerBookInfo = require("apps/filemanager/filemanagerbookinfo")
 local FileManagerCollection = require("apps/filemanager/filemanagercollection")
+local FileManagerDocument = require("document/filemanagerdocument")
```
I don't like that name (file and variable, and file in document/ but named as the other FileManagerModules in apps/filemanager/).
No good other idea though.

poire-z commented on frontend/apps/reader/readerui.lua:
```diff
@@ -559,9 +560,15 @@ function ReaderUI:showReader(file, provider)
+    -- `readerui:doShowReader` sent us here but there're some files better handled on the FM
+    provider = provider or DocumentRegistry:getProvider(file)
+    if provider.provider == "filemanagerdocument" then
+        self:showFileManager(file, FileManagerDocument.open)
```
Will it always be `FileManagerDocument.open` ? If yes, I would rather pass a boolean `is_filemanagerdocument=true` intsead of `filemanagerdocument_handle_callback`.

poire-z commented on frontend/document/filemanagerdocument.lua:
```diff
@@ -0,0 +1,85 @@
+end
+
+function FileManagerDocument:_readMetadata()
+    Document._readMetadata(self)
```
No issue with `Document:_readMetadata()` doing `self.info.number_of_pages = self._document:getPages() ` and your `_document = false,` ?

(I don't really like this `FileManagerDocument` - we need to worry at making look enough like a real document depending on what bit of code may still want to see it as a Document subclass - mostly just to get the registry stuff working... but well, no intuition about a better alternative, and I don't want to think too much :) )

pazos replied:
> No issue with Document:_readMetadata() doing self.info.number_of_pages = self._document:getPages() and your _document = false, ?

Not sure, at least it doesn't crash.

On coverbrowser it shows metadata: 0 pages. Hopefully we can delete that chunk of metadata or replace it with more useful info.

poire-z commented on plugins/shell-runner.koplugin/_meta.lua:
```diff
@@ -0,0 +1,6 @@
+local _ = require("gettext")
+return {
+    name = "shellrunner",
+    fullname = _("Shell Runner"),
+    description = _([[Execute scripts instead of view them]]),
```
So, which is better and which should be killed, between this as a plugin, and @NiLuJe 's "core" shell execution support ?

pazos replied:
the plugin is just to showcase the behaviour. It isn't intended to be merged (or even reviewed, outside the function that registers a handler)

pazos commented on frontend/document/filemanagerdocument.lua:
```diff
@@ -0,0 +1,85 @@
+    return true
+end
+
+function FileManagerDocument:getCoverPageImage()
+    return nil
```
Indeed, I can try to make it working now. As I said I'm not familiar enough with `Document`'s

Is it ok to return the path of the svg file here, in case there's one?

poire-z replied:
No, it is expected to return a bb.
And `RenderImage:renderSVGImageFile(filename, width, height, zoom)` can do just that.

pazos replied:
I'm afraid we need to hack `bookinfomanager.lua` a little bit for it to happen.

I say I'm afraid because I'm completely ignorant of the code there.

I can say just two things ATM:

in `BookInfoManager:getBookInfo` if I return early when provider.provider is a filemanagerdocument a bookinfo table with `ignore_cover = "Y"` everything works but there's no image.

If I comment `ignore_cover`  then `getBookInfo` is called again and again.

Attaching the diff of the code I'm playing with

```diff
diff --git a/plugins/coverbrowser.koplugin/bookinfomanager.lua b/plugins/coverbrowser.koplugin/bookinfomanager.lua
index 003f59eb..902bd16d 100644
--- a/plugins/coverbrowser.koplugin/bookinfomanager.lua
+++ b/plugins/coverbrowser.koplugin/bookinfomanager.lua
@@ -317,7 +317,7 @@ end
 -- Bookinfo management
 function BookInfoManager:getBookInfo(filepath, get_cover)
     local directory, filename = util.splitFilePathName(filepath)
-
+    logger.info("getBookInfo")
     -- CoverBrowser may be used by PathChooser, which will not filter out
     -- files with unknown book extension. If not a supported extension,
     -- returns a bookinfo like-object enough for a correct display and
@@ -344,6 +344,33 @@ function BookInfoManager:getBookInfo(filepath, get_cover)
         }
     end
 
+    if DocumentRegistry:getProvider(filepath).provider == "filemanagerdocument" then
+        logger.info("FM document returns early")
+        local bookinfo = {
+            directory = directory,
+            filename = filename,
+            in_progress = 0,
+            cover_fetched = "Y",
+            has_meta = nil,
+            has_cover = true,
+            ignore_meta = "Y",
+            --ignore_cover = "Y",
+            _is_directory = false,
+            _no_provider = false,
+        }
+
+        local icon = require("document/filemanagerdocument"):getIcon(filepath)
+        logger.info("file", filepath, "got icon", icon)
+
+        if icon then
+            local cover_bb = require("ui/renderimage"):renderSVGImageFile(icon, 100, 100, 1)
+            cover_bb:setAllocated(1)
+            bookinfo["cover_bb"] = cover_bb
+        end
+
+        logger.info("cover blitbuffer", bookinfo.cover_bb)
+        return bookinfo
+    end
     self:openDbConnection()
     local row = self.get_stmt:bind(directory, filename):step()
     -- NOTE: We do not reset right now because we'll be querying a BLOB,
@@ -397,6 +424,7 @@ function BookInfoManager:getBookInfo(filepath, get_cover)
 end
 
 function BookInfoManager:extractBookInfo(filepath, cover_specs)
+    logger.info("extractBookInfo", filepath)
     -- This will be run in a subprocess
     -- We use a temporary directory for cre cache (that will not affect parent process),
     -- so we don't fill the main cache with books we're not actually reading
```

poire-z replied:
There shouldn't be anything specific to hack in CoverBrowser.
The bit you took inspiration from is for directories.

You made your files be "FilemanagerDocument", so a document. They will go through `BookInfoManager:extractBookInfo(filepath)` where your FilemanagerDocument regular methods should be called:
```lua
DocumentRegistry:openDocument(filepath)
pages = document:getPageCount() -- remember I said this one may crash, may be the reason it did not work?
local props = document:getProps()
local cover_bb = document:getCoverPageImage()
document:close()
```
So, if you have defined those, they should be called to get (and cache) what's necessary to draw the coverbrowser items.

pazos replied:
thanks for the pointers.

we still need to hack bookinfomanager because the cover image cannot be derived from the filemanagerdocument without knowing which kind of extension the current `filename` has.

poire-z replied:
Probably the job of filemanagerdocument to find out what it's main provider is ? or find out in its docsetting what was the last/default provider use (if that is saved?) or in the global settings what the default provider for that extension is.

> cannot be derived from the filemanagerdocument without knowing which kind of extension the current filename has.

Document (and its derived ones) has a self.file member.

pazos replied:
Thanks!

> Document (and its derived ones) has a self.file member.

That's cool!

> Probably the job of filemanagerdocument to find out what it's main provider is ? or find out in its docsetting what was the last/default provider use (if that is saved?) or in the global settings what the default provider for that extension is.

The first intention was that. ie: store priority of handlers/providers internally and open stuff with the nº 1. But I was not sure because future usages would attempt to escalate priorities to be placed on top of the rest. So finally I removed priority stuff and decided it was easier and leaner to show a popup that lets the user choose between handlers.

Not sure if it does worth the effort to add priorities again just for the icons or, instead, just iterate over them and return the blitbuffer of the first svg for each extension.

pazos commented:
Thanks for the feedback. Much appreciated!

I just did some changes to address @poire-z concerns about changes in core and naming. Hopefully it is a bit better now :)
I also added a very basic way to prevent duplicates.

And simplified a bit the priority thing:

- avoid priority as a field of handlers
- register extensions with low priority on the documentRegistry.
- show a menu with possible actions if there're more than one.

poire-z commented:
> Hopefully it is a bit better now :)

Yep it is :) (5 lines added in ReaderUI, well done.)

NiLuJe commented on frontend/apps/reader/readerui.lua:
```diff
@@ -561,9 +561,16 @@ function ReaderUI:showReader(file, provider)
         return
     end
 
+    -- `readerui:doShowReader` sent us here but there're some files better handled on the FM
```
nit: `by` the FM ;)

pazos commented:
More things to do:

* Rename FileManagerDocument to FileDocument because they're not strictly related to the FM and currently they're handled on readerui. IMHO FileDocument is a better representation of the purpose: to do something with files.

* Prevent FileDocument's to be added to reader-centric places, such as history. Many others maybe, I just don't know yet.

poire-z commented:
> IMHO FileDocument is a better representation of the purpose: to do something with files.

But it's still not a document, it's a lot more a file than a document, it is not a document :)
And all real documents are also files.

So (for fun, if one manages to make us smile each time we see it):
`NotADocument`
`Focument` (_Faux_-Document)
`HandledFile`
`JustAFile`

pazos commented:
> > IMHO FileDocument is a better representation of the purpose: to do something with files.
> 
> But it's still not a document, it's a lot more a file than a document, it is not a document :) And all real documents are also files.
> 

It is a document :p. A supported document. The program knows how to handle it. All files that are not documents will lead to "file not supported" while a FileDocument will, by definition, never reach that.


> So (for fun, if one manages to make us smile each time we see it): `NotADocument` `Focument` (_Faux_-Document) `HandledFile` `JustAFile`

While I still believe it is a document I wouldn't mind calling it NotADocument. Not very readable as plain `notadocument.lua`. Maybe `NonDocument`?

NonDocument: a document without superpowers.

poire-z commented:
I'm super fine with NonDocument !

("KOReader is a document viewer". "multi-format documents: supports fixed page formats (PDF, DjVu, CBT, CBZ) and reflowable e-book formats (EPUB, FB2, Mobi, DOC, CHM, TXT)".
For me, a Document is something with textual or images content we read :) I don't consider executing shell scripts or extracting a zip file being "reading documents".)

pazos commented:
Makes sense to call it `NonDocument` if we don't want its children on the history of documents :)

pazos commented:
Some progress (or not :p) with icons.

That's what I get with `renderSVGImageFile`

![2022-10-14_15-00](https://user-images.githubusercontent.com/975883/195853309-7ba2d1c0-0e80-4d20-918d-55abf462d591.png)


A nice black box. 

It might be something obvious. Who knows ? :)

Frenzie commented:
What's the input?

On Oct 14 2022, at 3:03 pm, Martín Fernández ***@***.***> wrote:
>
>
> Some progress (or not :p) with icons.
> That's what I get with renderSVGImageFile
>
> A nice black box.
> It might be something obvious. Who knows ? :)
> —
> Reply to this email directly, view it on GitHub (https://github.com/koreader/koreader/pull/9615#issuecomment-1278982127), or unsubscribe (https://github.com/notifications/unsubscribe-auth/AABRQBP4RUGFR5SS7PUYHCTWDFKYVANCNFSM6AAAAAARA3NCDA).
> You are receiving this because you are subscribed to this thread.
>

poire-z commented:
> A nice black box.

A perfect icon for a NonDocument without superpowers !

RenderSVGImage returns a 2nd argument, is_straight_alpha, which may have it use different methods when blitting:
https://github.com/koreader/koreader/blob/d9eb6e971742810bb1ee545f80686a84b90c06d3/frontend/ui/widget/imagewidget.lua#L171-L174

https://github.com/koreader/koreader/blob/d9eb6e971742810bb1ee545f80686a84b90c06d3/frontend/ui/widget/imagewidget.lua#L211-L224

So, you may need to blit the initial bb (from renderSVGImage) onto a second bb so it's the alpha kind expected by all other codes.
Or you can try using `renderSVGImageFileWithMupdf()` instead of `renderSVGImageFile()`.

(I'm still not competent on alpha, @NiLuJe is our alpha-male.)

NiLuJe commented on frontend/document/nondocument.lua:
```diff
@@ -0,0 +1,129 @@
+        if v.icon_path then
+            logger.info("got image", v.icon_path)
+            local cover_bb = RenderImage:renderSVGImageFile(v.icon_path, 100, 100, 1)
+            cover_bb:setAllocated(1)
```
Redundant: RenderImage *always* returns newly allocated BBs ;).

NiLuJe commented:
> So, you may need to blit the initial bb (from renderSVGImage) onto a second bb so it's the alpha kind expected by all other codes.
> Or you can try using `renderSVGImageFileWithMupdf()` instead of `renderSVGImageFile()`.

Yeah, bookinfo doesn't forward the type of alpha (if any), which means we effectively don't support alpha *at all* for covers (not without code & DB changes, anyway), so you'll have to do a composite pass yourself to squash alpha.

pazos commented:
@NiLuJe @poire-z

Much better, thanks!

![2022-10-14_20-51](https://user-images.githubusercontent.com/975883/195920041-7cbd2864-3786-4bd7-8265-896b9945eda4.png)

If I understand correctly we want to have *alpha dealing* as part of `RenderImage:renderSVGImageFileWithNanoSVG`

Attaching the relevant changes. Bear with me. I, for the most part, don't really know what I'm doing :)

```diff
diff --git a/frontend/ui/renderimage.lua b/frontend/ui/renderimage.lua
index 70b660eb..3c0c29a9 100644
--- a/frontend/ui/renderimage.lua
+++ b/frontend/ui/renderimage.lua
@@ -6,6 +6,7 @@ All the render* functions should return a BlitBuffer object on success, nil othe

 local Blitbuffer = require("ffi/blitbuffer")
 local Math = require("optmath")
+local Screen = require("device").screen
 local ffi = require("ffi")
 local logger = require("logger")

@@ -369,13 +370,23 @@ function RenderImage:renderSVGImageFileWithNanoSVG(filename, width, height, zoom
         offset_y = Math.round((height - inner_h) / 2)
     end
     logger.dbg("renderSVG", filename, zoom, native_w, native_h, ">", width, height, offset_x, offset$
-    local bb = Blitbuffer.new(width, height, Blitbuffer.TYPE_BBRGB32)
-    svg_image:drawTo(bb, zoom, offset_x, offset_y)
+    local file_bb = Blitbuffer.new(width, height, Blitbuffer.TYPE_BBRGB32)
+    svg_image:drawTo(file_bb, zoom, offset_x, offset_y)
     svg_image:free()
+
+    -- handle alpha
+    local bb = Blitbuffer.new(width, height, Blitbuffer.TYPE_BBRGB32)
+    bb:fill(Blitbuffer.COLOR_WHITE)
+    if Screen.sw_dithering then
+        bb:ditheralphablitFrom(file_bb, 0, 0, 0, 0, bb.w, bb.h)
+    else
+        bb:alphablitFrom(file_bb, 0, 0, 0, 0, bb.w, bb.h)
+    end
     return bb, true -- is_straight_alpha=true
 end

```

NiLuJe commented:
> If I understand correctly we want to have _alpha dealing_ as part of `RenderImage:renderSVGImageFileWithNanoSVG`

No we don't ;o).

There are many and varied use-cases where we *can* and *do* honor alpha, that's left to the caller.

The issue here is that the caller doesn't actually *display* the BB right away, it instead goes through bookinfo, which does not keep track of whether alpha matters to the bb.

So, here, there are only two choices:

* Either update bookinfo & the db to keep track of an dhonor both types of alpha properly
* Do a composition pass in your caller (hint: pick this ;p).

NiLuJe commented:
> * Do a composition pass in your caller (hint: pick this ;p).

Or something like you proposed, but behind an argument check, because we do *not* want to dumb down ReaderImage by default ;).

poire-z commented:
> (But my vote is definitely to leave it to the caller, FWIW).

My vote too.
Do what you added into renderSVGImageFileWithNanoSVG() in  your `NonDocument:getCoverPageImage()` depending on the 2nd returned value you'll get from it.

(I don't think there is any issue if we get a SVG image from a credocument:getCoverImage(), as not knowing its type, we call `RenderImage:renderImageData()` which itself will call `renderSVGImageDataWithCRengine()` which should do the right thing.
But for things we know are simple SVG, it's saner to not load crengine and to use renderSVGImageFileWithNanoSVG() and make it a normal bb by drawing it on another filled bb.)

pazos commented on frontend/document/nondocument.lua:
```diff
@@ -0,0 +1,165 @@
+    local extension = util.getFileNameSuffix(self.file)
+    for __, v in ipairs(self.extensions[extension]) do
+        if v.icon_path then
+            return getIcon(v.icon_path, 100, 100, 4)
```
@poire-z ·@NiLuJe: any magic trick to put sensible values here for width, height and zoom factor?

I've got current values by trial and error and, even if they work on **my** emulator I'm not sure that's good enough.

poire-z replied:
No obvious idea.
Even if it's not really an icon, and it may get big in Mosaic view, just do as in:
https://github.com/koreader/koreader/blob/741562e84c2a33e38858ef379b3db95406a75a79/frontend/ui/widget/iconwidget.lua#L39-L40
and zoom nil.

pazos commented:
I'm thinking about moving the examples to the hello plugin, so they can stay there for reference.

pazos commented:
> No obvious idea. Even if it's not really an icon, and it may get big in Mosaic view, just do as in:
> 
> https://github.com/koreader/koreader/blob/741562e84c2a33e38858ef379b3db95406a75a79/frontend/ui/widget/iconwidget.lua#L39-L40
> 
> 
> and zoom nil.

I just tried an the result is not good.
![2022-10-17_01-47](https://user-images.githubusercontent.com/975883/196791975-3fb41927-55f0-46bb-a89a-de9014755c44.png)

Left icon is following the suggestion. Looks pixelated. Right icon is with size * 4. Looks good. (any multiplier of the size makes the image look better but 4 looks like the sweet spot)

Also, icons look to big, so I aded a bit of margin: Both icons here are with the trick `size * 4`.

![2022-10-19_21-56](https://user-images.githubusercontent.com/975883/196792013-e2c39247-e07a-4fad-9632-e8c7b8bf58e8.png)

poire-z commented:
You could try with `w = h = min(Screen.width, Screen.height) / 3` to aim at 3x3 cover images. It could then be scaled down (but still look ok?) to be adjusted to the displayed size and saved in coverbrowser cache.

pazos commented:
> You could try with w = h = min(Screen.width, Screen.height) / 3 to aim at 3x3 cover images. It could then be scaled down (but still look ok?) to be adjusted to the displayed size and saved in coverbrowser cache.

Great, thanks!

Moved the example to `hello.koplugin`. 

Here's a sneak peek

Mosaic mode:

![2022-10-20_12-49](https://user-images.githubusercontent.com/975883/196929731-80571bbc-b2df-4766-9e8a-b04e0a9f74ba.png)

Same files in list mode, showing the action list.
![Peek 2022-10-20 12-48](https://user-images.githubusercontent.com/975883/196929959-621c2482-e076-4fa2-bda3-bab66ef4b445.gif)

yparitcher commented on plugins/hello.koplugin/main.lua:
```diff
@@ -1,53 +1,98 @@
-end
+-- This plugin is disabled by default as it has no real usage for end users.
+-- Remove the following if block to enable it
+--if true then
+--    return { disabled = true, }
+--end
```
Remember to uncomment this :)

pazos replied:
I will. Thanks for the reminder.

pazos commented:
Lets bump this PR :)

I made a nice icon for a KOReader package.

![Captura de pantalla_2023-10-15_01-17-44](https://github.com/koreader/koreader/assets/975883/24971654-ea01-4123-9653-781d13f879c7)
![Captura de pantalla_2023-10-15_01-18-39](https://github.com/koreader/koreader/assets/975883/b6bc0d13-5b91-44ef-a913-a67a81999545)

As you can see it is a square.

I'm wondering if we should impose the image ratio here, since plugins can include their own svgs and displaying those as is could be ugly.

Frenzie commented:
Wouldn't an author come to the same conclusion organically if they deviated too far from 3:2?

poire-z commented:
> As you can see it is a square.

So, looks fine to me: it's not a book, it's a package, it's a cube, it feels alright that it is displayed in a square :)

> I'm wondering if we should impose the image w/h ratio here, since plugins can include their own svgs and displaying those as is could be ugly.

Would have to be handled in CoverBrowser, which knows the max h/w available for thumbnails, and would have to do something or provide it to the cover image fetching or override it.
Postpone that to later ? (until you get used to seeing squares and non-square would feel then odd :) a square feels alright to distinguish these from books/documents).

pazos commented:
Superseeded by #11056
